app/services/structured_fact_importer.rb

```diff
@@ -0,0 +1,37 @@
+class StructuredFactImporter < FactImporter
+  def normalize(facts)
+    # Remove empty values first, so nil facts added by normalize_recurse imply compose
+    facts = facts.select { |k, v| v.present? }
+    normalize_recurse({}, facts)
+  end
+
+  # expand {'a' => {'b' => 'c'}} to {'a' => nil, 'a::b' => 'c'}
```

Comment:
Yes, I did see the code in the plugins and took inspiration from it, though this does work quite differently (it should be simpler) from what I could understand of the plugins.

Plugins should simply be able to inherit from StructuredFactImporter rather than using this method directly. This class should automatically handle everything if the input is a hash of hashes. If the input is a hash of separator-based fact names then plugins should override `normalize` to call that original `unsparse` implementation first, then the rest should just work.